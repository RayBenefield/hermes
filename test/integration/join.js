import flame from '@leonardvandriel/flame';
import * as databases from './database'; // eslint-disable-line import/no-unresolved,import/extensions,no-unused-vars
import uuid from '../utils/mock-uuid';
import random from '../utils/mock-random';
import configureSave from '../utils/save-db';
import configureDomain from '../../src/domain';
import configureFlame from '../../src/lib/local-db';

const db = configureFlame(flame);
const domain = configureDomain({ db, uuid, random });
const save = configureSave(flame); // eslint-disable-line no-unused-vars
beforeEach(() => flame.loadDatabase(undefined));

test('player does not exist', () => {
    // Given
    const lead = {
        platform: 'facebook',
        id: 123456,
    };

    // When
    return domain({
        lead,
        action: 'join',
    })
        .then(results => expect({
            results,
            db: flame.get('/'),
        }).toMatchSnapshot());
});

test('add player to the queue', () => {
    // Given
    const lead = {
        platform: 'facebook',
        id: 123456,
    };
    flame.loadDatabase(databases['existing-player']);

    // When
    return domain({
        lead,
        action: 'join',
    })
        .then(save('player-in-queue'))
        .then(results => expect({
            results,
            db: flame.get('/'),
        }).toMatchSnapshot());
});

test('player already added to the queue', () => {
    // Given
    const lead = {
        platform: 'facebook',
        id: 123456,
    };
    flame.loadDatabase(databases['player-in-queue']);

    // When
    return domain({
        lead,
        action: 'join',
    })
        .then(results => expect({
            results,
            db: flame.get('/'),
        }).toMatchSnapshot());
});

test('new game started from full queue', () => {
    // Given
    const lead = {
        platform: 'facebook',
        id: 123456,
    };
    flame.loadDatabase(databases['4-players-in-queue']);

    // When
    return domain({
        lead,
        action: 'join',
    })
        .then(results => expect({
            results,
            db: flame.get('/'),
        }).toMatchSnapshot());
});
