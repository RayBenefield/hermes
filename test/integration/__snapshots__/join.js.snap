// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add fourth player to the queue 1`] = `
Object {
  "db": Object {
    "players": Object {
      "facebook": Object {
        "123456": Object {
          "first_name": "Ray",
          "id": 123456,
          "platform": "facebook",
        },
        "234567": Object {
          "first_name": "Hermon",
          "id": 234567,
          "platform": "facebook",
        },
        "345678": Object {
          "first_name": "Hermes",
          "id": 345678,
          "platform": "facebook",
        },
        "456789": Object {
          "first_name": "Hermshire",
          "id": 456789,
          "platform": "facebook",
        },
        "567890": Object {
          "first_name": "Hermda",
          "id": 567890,
          "platform": "facebook",
        },
      },
    },
    "queue": Object {
      "123456": true,
      "234567": true,
      "345678": true,
      "456789": true,
    },
  },
  "results": Object {
    "action": "join",
    "lead": Object {
      "id": 456789,
      "platform": "facebook",
    },
    "messages": Array [
      Object {
        "payload": Object {
          "player": 456789,
          "queue": Array [
            "123456",
            "234567",
            "345678",
            456789,
          ],
        },
        "type": "queue-joined-message",
      },
    ],
    "player": Object {
      "first_name": "Hermshire",
      "id": 456789,
      "platform": "facebook",
    },
    "players": Array [
      Object {
        "first_name": "Ray",
        "id": 123456,
        "platform": "facebook",
      },
      Object {
        "first_name": "Hermon",
        "id": 234567,
        "platform": "facebook",
      },
      Object {
        "first_name": "Hermes",
        "id": 345678,
        "platform": "facebook",
      },
    ],
    "queue": Array [
      "123456",
      "234567",
      "345678",
    ],
  },
}
`;

exports[`add player to the queue 1`] = `
Object {
  "db": Object {
    "players": Object {
      "facebook": Object {
        "123456": Object {
          "first_name": "Ray",
          "id": 123456,
          "platform": "facebook",
        },
        "234567": Object {
          "first_name": "Hermon",
          "id": 234567,
          "platform": "facebook",
        },
        "345678": Object {
          "first_name": "Hermes",
          "id": 345678,
          "platform": "facebook",
        },
        "456789": Object {
          "first_name": "Hermshire",
          "id": 456789,
          "platform": "facebook",
        },
        "567890": Object {
          "first_name": "Hermda",
          "id": 567890,
          "platform": "facebook",
        },
      },
    },
    "queue": Object {
      "123456": true,
    },
  },
  "results": Object {
    "action": "join",
    "lead": Object {
      "id": 123456,
      "platform": "facebook",
    },
    "messages": Array [
      Object {
        "payload": Object {
          "player": 123456,
          "queue": Array [
            123456,
          ],
        },
        "type": "queue-joined-message",
      },
    ],
    "player": Object {
      "first_name": "Ray",
      "id": 123456,
      "platform": "facebook",
    },
    "players": Array [],
    "queue": Array [],
  },
}
`;

exports[`add second player to the queue 1`] = `
Object {
  "db": Object {
    "players": Object {
      "facebook": Object {
        "123456": Object {
          "first_name": "Ray",
          "id": 123456,
          "platform": "facebook",
        },
        "234567": Object {
          "first_name": "Hermon",
          "id": 234567,
          "platform": "facebook",
        },
        "345678": Object {
          "first_name": "Hermes",
          "id": 345678,
          "platform": "facebook",
        },
        "456789": Object {
          "first_name": "Hermshire",
          "id": 456789,
          "platform": "facebook",
        },
        "567890": Object {
          "first_name": "Hermda",
          "id": 567890,
          "platform": "facebook",
        },
      },
    },
    "queue": Object {
      "123456": true,
      "234567": true,
    },
  },
  "results": Object {
    "action": "join",
    "lead": Object {
      "id": 234567,
      "platform": "facebook",
    },
    "messages": Array [
      Object {
        "payload": Object {
          "player": 234567,
          "queue": Array [
            "123456",
            234567,
          ],
        },
        "type": "queue-joined-message",
      },
    ],
    "player": Object {
      "first_name": "Hermon",
      "id": 234567,
      "platform": "facebook",
    },
    "players": Array [
      Object {
        "first_name": "Ray",
        "id": 123456,
        "platform": "facebook",
      },
    ],
    "queue": Array [
      "123456",
    ],
  },
}
`;

exports[`add third player to the queue 1`] = `
Object {
  "db": Object {
    "players": Object {
      "facebook": Object {
        "123456": Object {
          "first_name": "Ray",
          "id": 123456,
          "platform": "facebook",
        },
        "234567": Object {
          "first_name": "Hermon",
          "id": 234567,
          "platform": "facebook",
        },
        "345678": Object {
          "first_name": "Hermes",
          "id": 345678,
          "platform": "facebook",
        },
        "456789": Object {
          "first_name": "Hermshire",
          "id": 456789,
          "platform": "facebook",
        },
        "567890": Object {
          "first_name": "Hermda",
          "id": 567890,
          "platform": "facebook",
        },
      },
    },
    "queue": Object {
      "123456": true,
      "234567": true,
      "345678": true,
    },
  },
  "results": Object {
    "action": "join",
    "lead": Object {
      "id": 345678,
      "platform": "facebook",
    },
    "messages": Array [
      Object {
        "payload": Object {
          "player": 345678,
          "queue": Array [
            "123456",
            "234567",
            345678,
          ],
        },
        "type": "queue-joined-message",
      },
    ],
    "player": Object {
      "first_name": "Hermes",
      "id": 345678,
      "platform": "facebook",
    },
    "players": Array [
      Object {
        "first_name": "Ray",
        "id": 123456,
        "platform": "facebook",
      },
      Object {
        "first_name": "Hermon",
        "id": 234567,
        "platform": "facebook",
      },
    ],
    "queue": Array [
      "123456",
      "234567",
    ],
  },
}
`;

exports[`new game started from full queue 1`] = `
Object {
  "db": Object {
    "games": Object {
      "0-05526803409810505": Object {
        "id": "0-05526803409810505",
        "notified_players": Object {
          "567890": true,
        },
        "players": Object {
          "123456": true,
          "234567": true,
          "345678": true,
          "456789": true,
          "567890": true,
        },
      },
    },
    "players": Object {
      "facebook": Object {
        "123456": Object {
          "first_name": "Ray",
          "id": 123456,
          "platform": "facebook",
        },
        "234567": Object {
          "first_name": "Hermon",
          "id": 234567,
          "platform": "facebook",
        },
        "345678": Object {
          "first_name": "Hermes",
          "id": 345678,
          "platform": "facebook",
        },
        "456789": Object {
          "first_name": "Hermshire",
          "id": 456789,
          "platform": "facebook",
        },
        "567890": Object {
          "first_name": "Hermda",
          "id": 567890,
          "platform": "facebook",
        },
      },
    },
    "queue": Object {},
  },
  "results": Object {
    "action": "join",
    "lead": Object {
      "id": 567890,
      "platform": "facebook",
    },
    "messages": Array [
      Object {
        "payload": Object {
          "players": Array [
            Object {
              "first_name": "Hermda",
              "id": 567890,
            },
            Object {
              "first_name": "Ray",
              "id": 123456,
            },
            Object {
              "first_name": "Hermon",
              "id": 234567,
            },
            Object {
              "first_name": "Hermes",
              "id": 345678,
            },
            Object {
              "first_name": "Hermshire",
              "id": 456789,
            },
          ],
        },
        "type": "game-started-message",
      },
    ],
    "player": Object {
      "first_name": "Hermda",
      "id": 567890,
      "platform": "facebook",
    },
    "players": Array [
      Object {
        "first_name": "Ray",
        "id": 123456,
        "platform": "facebook",
      },
      Object {
        "first_name": "Hermon",
        "id": 234567,
        "platform": "facebook",
      },
      Object {
        "first_name": "Hermes",
        "id": 345678,
        "platform": "facebook",
      },
      Object {
        "first_name": "Hermshire",
        "id": 456789,
        "platform": "facebook",
      },
    ],
    "queue": Array [
      "123456",
      "234567",
      "345678",
      "456789",
    ],
  },
}
`;

exports[`player already added to the queue 1`] = `
Object {
  "db": Object {
    "players": Object {
      "facebook": Object {
        "123456": Object {
          "id": 123456,
          "platform": "facebook",
        },
      },
    },
    "queue": Object {
      "123456": true,
    },
  },
  "results": Object {
    "action": "join",
    "lead": Object {
      "id": 123456,
      "platform": "facebook",
    },
    "messages": Array [
      Object {
        "payload": Object {
          "queue": Array [
            "123456",
          ],
        },
        "type": "already-in-queue-message",
      },
    ],
    "player": Object {
      "id": 123456,
      "platform": "facebook",
    },
    "players": Array [
      Object {
        "id": 123456,
        "platform": "facebook",
      },
    ],
    "queue": Array [
      "123456",
    ],
  },
}
`;

exports[`player does not exist 1`] = `
Object {
  "db": null,
  "results": Object {
    "action": "join",
    "lead": Object {
      "id": 123456,
      "platform": "facebook",
    },
    "messages": Array [
      Object {
        "type": "player-does-not-exist-message",
      },
    ],
    "player": undefined,
    "players": Array [],
    "queue": Array [],
  },
}
`;
